### PR TITLE
Optimize discount calculations

### PR DESCRIPTION
**Pull request by patrys**
This change makes sure that all discount information is prefetched ahead of the time.

### Before

```
DEBUG django.db.backends (0.001) SELECT "discount_sale"."id", "discount_sale"."name", "discount_sale"."type", "discount_sale"."value", "discount_sale"."start_date", "discount_sale"."end_date" FROM "discount_sale" WHERE (("discount_sale"."end_date" IS NULL OR "discount_sale"."end_date" >= '2019-06-17'::date) AND "discount_sale"."start_date" <= '2019-06-17'::date); args=(datetime.date(2019, 6, 17), datetime.date(2019, 6, 17)) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.042) SELECT ("discount_sale_products"."sale_id") AS "_prefetch_related_val_sale_id", "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."tax_rate", "product_product"."weight" FROM "product_product" INNER JOIN "discount_sale_products" ON ("product_product"."id" = "discount_sale_products"."product_id") WHERE "discount_sale_products"."sale_id" IN (1, 2, 3, 4, 5) ORDER BY "product_product"."name" ASC; args=(1, 2, 3, 4, 5) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.002) SELECT ("discount_sale_categories"."sale_id") AS "_prefetch_related_val_sale_id", "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" INNER JOIN "discount_sale_categories" ON ("product_category"."id" = "discount_sale_categories"."category_id") WHERE "discount_sale_categories"."sale_id" IN (1, 2, 3, 4, 5); args=(1, 2, 3, 4, 5) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT ("discount_sale_collections"."sale_id") AS "_prefetch_related_val_sale_id", "product_collection"."id", "product_collection"."publication_date", "product_collection"."is_published", "product_collection"."seo_title", "product_collection"."seo_description", "product_collection"."name", "product_collection"."slug", "product_collection"."background_image", "product_collection"."background_image_alt", "product_collection"."description", "product_collection"."description_json" FROM "product_collection" INNER JOIN "discount_sale_collections" ON ("product_collection"."id" = "discount_sale_collections"."collection_id") WHERE "discount_sale_collections"."sale_id" IN (1, 2, 3, 4, 5) ORDER BY "product_collection"."slug" ASC; args=(1, 2, 3, 4, 5) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 7; args=(7,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 7; args=(7,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 7; args=(7,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 7; args=(7,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 8; args=(8,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id", "product_category"."seo_title", "product_category"."seo_description", "product_category"."name", "product_category"."slug", "product_category"."description", "product_category"."description_json", "product_category"."parent_id", "product_category"."background_image", "product_category"."background_image_alt", "product_category"."lft", "product_category"."rght", "product_category"."tree_id", "product_category"."level" FROM "product_category" WHERE "product_category"."id" = 9; args=(9,) [PID:4120:Thread-1]
```

### After

```
DEBUG django.db.backends (0.001) SELECT "discount_sale"."id", "discount_sale"."name", "discount_sale"."type", "discount_sale"."value", "discount_sale"."start_date", "discount_sale"."end_date" FROM "discount_sale" WHERE (("discount_sale"."end_date" IS NULL OR "discount_sale"."end_date" >= '2019-06-17'::date) AND "discount_sale"."start_date" <= '2019-06-17'::date); args=(datetime.date(2019, 6, 17), datetime.date(2019, 6, 17)) [PID:5057:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "discount_sale_collections"."sale_id", "discount_sale_collections"."collection_id" FROM "discount_sale_collections" WHERE "discount_sale_collections"."sale_id" IN (1, 2, 3, 4, 5); args=(1, 2, 3, 4, 5) [PID:5057:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "discount_sale_products"."sale_id", "discount_sale_products"."product_id" FROM "discount_sale_products" WHERE "discount_sale_products"."sale_id" IN (1, 2, 3, 4, 5); args=(1, 2, 3, 4, 5) [PID:5057:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "discount_sale_categories"."sale_id", "discount_sale_categories"."category_id" FROM "discount_sale_categories" WHERE "discount_sale_categories"."sale_id" IN (1, 2, 3, 4, 5); args=(1, 2, 3, 4, 5) [PID:5057:Thread-1]
```
```
DEBUG django.db.backends (0.001) SELECT "product_category"."id", "product_category"."parent_id", "product_category"."lft", "product_category"."rght", "product_category"."tree_id" FROM "product_category" WHERE "product_category"."id" IN (8, 9, 7) ORDER BY "product_category"."tree_id" ASC, "product_category"."parent_id" ASC, "product_category"."lft" ASC; args=(8, 9, 7) [PID:5057:Thread-1]
```
```
DEBUG django.db.backends (0.000) SELECT "product_category"."id" FROM "product_category" WHERE (("product_category"."lft" >= 1 AND "product_category"."rght" <= 2 AND "product_category"."tree_id" = 1) OR ("product_category"."lft" >= 1 AND "product_category"."rght" <= 2 AND "product_category"."tree_id" = 2) OR ("product_category"."lft" >= 1 AND "product_category"."rght" <= 2 AND "product_category"."tree_id" = 3)) ORDER BY "product_category"."tree_id" ASC, "product_category"."lft" ASC; args=(1, 2, 1, 1, 2, 2, 1, 2, 3) [PID:5057:Thread-1]
```

### Query
```graphql
query ProductsList {
  products(first: 20) {
    edges {
      node {
        pricing {
          priceRange {
            start {
              gross {
                currency
                amount
                localized
              }
            }
          }
        }
      }
    }
  }
}

```

### Pull Request Checklist

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [ ] Database migration files are up to date.
1. [ ] The changes are tested.
1. [ ] The code is documented (docstrings, project documentation).
1. [ ] GraphQL schema and type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by django-queries** [comment]
Here is the report for 89bfda69dbfb9406593973f1cebe560d6c046fa2 (patrys/saleor @ optimize-discounts)
Base comparison is 7eded6db55abaf4be55ec56a97d4654c62a3b41f.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  add billing address to checkout     	         41	         41
  add shipping to checkout            	          7	          7
  checkout payment charge             	         16	         16
  complete checkout                   	          6	          6
  create checkout                     	         45	         45

# api.benchmark homepage
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve main menu                  	          5	          5
  retrieve product list               	          4	          4
  retrieve secondary menu             	          5	          5
  retrieve shop                       	          2	          2

# api.benchmark product
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  product details                     	         16	         16

# api.benchmark variant
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve variant list               	         13	         13
```

</p>
</details>

**Comment by NyanKiyoshi** [comment]
Might want to add this to the query count tests?

File: saleor/discount/utils.py [hunk 1]
```diff
@@ -17,37 +20,28 @@ def decrease_voucher_usage(voucher):
...
-    )
+def are_product_collections_on_sale(product, discount: DiscountInfo):
+    """Check if any collection is on sale."""
+    discounted_collections = discount.collection_ids
+    product_collections = set(c.id for c in product.collections.all())
+    return product_collections.intersection(discounted_collections)
 
 
-def get_product_discount_on_sale(sale, product):
+def get_product_discount_on_sale(discount: DiscountInfo, product):
```
Comment (derenio) [inline saleor/discount/utils.py]: For consistency we should place the `discount` argument as the second one.

Comment (patrys) [reply saleor/discount/utils.py]: We should but this is an unrelated change.

**Pull request merged**